File: sdks/typescript/src/http-errors.ts
/**
 * Parse FastAPI error bodies into SDK exceptions.
 */

import { QueryError, QueryOutOfScopeError, ServerError } from './errors.js';

const MAX_SUGGESTION_CHARS = 200;

function parseSuggestedQueries(raw: unknown): string[] {
  if (!Array.isArray(raw)) return [];
  const cleaned: string[] = [];
  for (const item of raw) {
    if (typeof item !== 'string') continue;
    const text = item.trim();
    if (!text) continue;
    cleaned.push(text.slice(0, MAX_SUGGESTION_CHARS));
    if (cleaned.length >= 3) break;
  }
  return cleaned;
}

function formatOutOfScopeMessage(reason: string, suggestedQueries: string[]): string {
  const reasonSuffix = reason ? ` (${reason})` : '';
  const hint =
    suggestedQueries.length > 0
      ? `. Try: ${suggestedQueries.map((q) => `"${q}"`).join(' · ')}`
      : '';
  return `Query out of scope${reasonSuffix}${hint}`;
}

function detailToMessage(detail: unknown): string {
  if (typeof detail === 'string') return detail;
  if (detail && typeof detail === 'object' && !Array.isArray(detail)) {
    const obj = detail as Record<string, unknown>;
    if (typeof obj.message === 'string' && obj.message.trim()) return obj.message;
    if (obj.detail === 'query_out_of_scope') {
      const reason = typeof obj.reason === 'string' ? obj.reason : '';
      const suggested = parseSuggestedQueries(obj.suggested_queries);
      return formatOutOfScopeMessage(reason, suggested);
    }
    if (obj.code === 'session_database_id_mismatch') {
      return (
        `Session is pinned to database "${obj.pinned_database_id ?? '?'}" ` +
        `but request used "${obj.requested_database_id ?? '?'}". ` +
        'Start a new chat session or select the pinned database.'
      );
    }
    if (typeof obj.code === 'string') return obj.code;
  }
  return String(detail);
}

export function raiseForResponse(status: number, body: { detail?: unknown }): never {
  const detail = body.detail ?? 'Request failed';

  if (status >= 500) {
    throw new ServerError(`Server error (${status}): ${detailToMessage(detail)}`, status);
  }

  if (
    status >= 400 &&
    detail &&
    typeof detail === 'object' &&
    !Array.isArray(detail) &&
    (detail as Record<string, unknown>).detail === 'query_out_of_scope'
  ) {
    const obj = detail as Record<string, unknown>;
    const reason = typeof obj.reason === 'string' ? obj.reason : '';
    const suggestedQueries = parseSuggestedQueries(obj.suggested_queries);
    const message = formatOutOfScopeMessage(reason, suggestedQueries);
    throw new QueryOutOfScopeError(message, status, reason, suggestedQueries);
  }

  throw new QueryError(`Query rejected (${status}): ${detailToMessage(detail)}`, status);
}

File: sdks/python/seal/_http_errors.py
"""Parse FastAPI error bodies into SDK exceptions."""

from __future__ import annotations

from typing import Any

from seal.exceptions import QueryError, QueryOutOfScopeError, ServerError
from seal.rate_limit import is_rate_limit_http, rate_limit_user_message

_SESSION_DATABASE_ID_MISMATCH = "session_database_id_mismatch"

RATE_LIMIT_USER_MESSAGE = rate_limit_user_message()


def is_rate_limit_signal(status: int, text: str) -> bool:
    """True when an HTTP status or message body indicates LLM rate limiting."""
    return is_rate_limit_http(status, text)


def _format_out_of_scope_message(*, reason: str, suggested_queries: list[str]) -> str:
    message = f"Query out of scope ({reason})" if reason else "Query out of scope"
    if suggested_queries:
        examples = " · ".join(f'"{q}"' for q in suggested_queries)
        message = f"{message}. Try: {examples}"
    return message


def _parse_suggested_queries(raw: Any) -> list[str]:
    if not isinstance(raw, list):
        return []
    cleaned: list[str] = []
    for item in raw:
        if isinstance(item, str) and item.strip():
            cleaned.append(item.strip()[:200])
        if len(cleaned) >= 3:
            break
    return cleaned


def _validation_item_message(item: Any) -> str:
    if isinstance(item, str):
        return item.strip()
    if isinstance(item, dict):
        for key in ("msg", "message", "detail"):
            value = item.get(key)
            if isinstance(value, str) and value.strip():
                return value.strip()
    return str(item).strip()


def detail_to_message(detail: Any) -> str:
    """Human-readable message from a FastAPI ``detail`` field (string or object)."""
    if isinstance(detail, str):
        return detail
    if isinstance(detail, list):
        parts = [_validation_item_message(item) for item in detail]
        joined = "; ".join(part for part in parts if part)
        return joined or str(detail)
    if isinstance(detail, dict):
        message = detail.get("message")
        if isinstance(message, str) and message.strip():
            return message
        if detail.get("detail") == "query_out_of_scope":
            reason = str(detail.get("reason") or "")
            suggested = _parse_suggested_queries(detail.get("suggested_queries"))
            return _format_out_of_scope_message(reason=reason, suggested_queries=suggested)
        code = detail.get("code")
        if code == _SESSION_DATABASE_ID_MISMATCH:
            pinned = detail.get("pinned_database_id", "?")
            requested = detail.get("requested_database_id", "?")
            return (
                f"Session is pinned to database {pinned!r} but request used {requested!r}. "
                "Start a new chat session or use the pinned database."
            )
        if isinstance(code, str):
            return code
    return str(detail)


def raise_for_response(response_status: int, detail: Any) -> None:
    """Raise an appropriate SDK exception for a non-2xx HTTP response."""
    if response_status >= 500:
        message = detail_to_message(detail)
        user_message = (
            RATE_LIMIT_USER_MESSAGE if is_rate_limit_signal(response_status, message) else message
        )
        raise ServerError(
            f"Server error ({response_status}): {user_message}",
            status_code=response_status,
        )

    if response_status >= 400:
        if isinstance(detail, dict) and detail.get("detail") == "query_out_of_scope":
            reason = str(detail.get("reason") or "")
            suggested = _parse_suggested_queries(detail.get("suggested_queries"))
            message = _format_out_of_scope_message(reason=reason, suggested_queries=suggested)
            raise QueryOutOfScopeError(
                message,
                reason=reason,
                suggested_queries=suggested,
                status_code=response_status,
            )
        message = detail_to_message(detail)
        raise QueryError(
            f"Query rejected ({response_status}): {message}",
            status_code=response_status,
        )
